client/src/main/java/io/pravega/schemaregistry/client/SchemaRegistryClient.java

```diff
@@ -0,0 +1,367 @@
+import io.pravega.schemaregistry.contract.data.SchemaInfo;
+import io.pravega.schemaregistry.contract.data.SchemaWithVersion;
+import io.pravega.schemaregistry.contract.data.VersionInfo;
+
+import javax.annotation.Nullable;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static io.pravega.schemaregistry.client.exceptions.RegistryExceptions.*;
+
+/**
+ * Defines a registry client for interacting with schema registry service. 
+ * The implementation of this interface should provide atomicity and read-after-write-consistency guarantees for all the methods.
+ */
+@Beta
+public interface SchemaRegistryClient {
+    /**
+     * Adds a new group. A group refers to the name under which the schemas are registered. A group is identified by a 
+     * unique id and has an associated set of group metadata {@link GroupProperties} and a list of codec types and a 
+     * versioned history of schemas that were registered under the group. 
+     * Add group is idempotent. If the group by the same id already exists the api will return false. 
+     * 
+     * @param groupId Id for the group that uniquely identifies the group. 
+     * @param groupProperties groupProperties Group properties for the group. These include serialization format, compatibility policy, 
+     *                        and flag to declare whether multiple schemas representing distinct object types can be 
+     *                        registered with the group. Type identify objects of same type. Schema compatibility checks 
+     *                        are always performed for schemas that share same {@link SchemaInfo#type}.
+     *                        Additionally, a user defined map of properties can be supplied.
+     * @return True indicates if the group was added successfully, false if it exists. 
+     * @throws BadArgumentException if the group properties is rejected by service.
+     * @throws UnauthorizedException if the user is unauthorized.
+     */
+    boolean addGroup(String groupId, GroupProperties groupProperties) throws BadArgumentException, UnauthorizedException;
+    
+    /**
+     * Removes a group identified by the groupId. This will remove all the codec types and schemas registered under the group.
+     * Remove group is idempotent. 
+     * 
+     * @param groupId Id for the group that uniquely identifies the group. 
+     * @throws UnauthorizedException if the user is unauthorized.
+     */
+    void removeGroup(String groupId) throws UnauthorizedException;
+
+    /**
+     * List all groups that the user is authorized on. This returns an iterator where each element is a pair of group 
+     * name and group properties. 
+     * This iterator can be used to iterate over each element until all elements are exhausted. 
+     * The implementation should guarantee that all groups added before and till the iterator continues to return 
+     * {@link Iterator#hasNext()} = true should be available for iteration. 
```

Comment:
2nd is what the implementation promises -- and the implementation detail is -- it makes network calls to service to get next set of groups from the continuation token.

Reply:
> 2nd is what the implementation promises 

Currently the implementation only delivers on the first. There is a race after the last reply from the server but before the user's final hasNext call, an item could be added and it would violate the above semantics.

I don't see an simple way to fix the impl, so I think it may be best to just update the javadoc to reflect that.

Reply:
> Currently the implementation only delivers on the first. There is a race after the last reply from the server but before the user's final hasNext call, an item could be added and it would violate the above semantics.

The exit is in the has next call. After having exhausted all retrieved elements, we make a call to service before returning false from `hasNext`. and if this is empty list, hasNext returns false. 
So there isnt a race of the kind you speak of